### DISCUSSION TITLE
Is it possible to setup a dynamic expiry check for each entry?

### DISCUSSION THREAD

**Discussion by SidB3** [ben-manes/caffeine#1786] [Q&A] [answered]
I have a`LoadingCache` that is backing a database. I don't want to expire an entry if the database entry hasn't changed.

Is it possible that at expiration time I can make a database call compare the entry and if they are unchanged then I extend the expiry time?

**Comment by ben-manes** [comment, marked as answer]
This is where `refreshAfterWrite` is useful. If the entry is accessed after that duration, but before expiration, then an async refresh is performed. The `CacheLoader.reload(key, oldValue)` can load the new value, optionally returning the old one is desired. This is used to hide latency rather than allow popular entries to expire and cause requests to stall waiting for it to be loaded anew.

If you do want to resurrect at eviction time then this can be achieved in a roundabout manner using a listener that reinserts the entry. The `Caffeine.removalListener` is called after the operation completes, so it could reinsert with a short race if a concurrent load happens concurrently. That minor problem could be mitigated using `Caffeine.evictionListener`, which is called within the atomic delete operation, to stage the entry into a temporary victim hashmap for an async reinsert or concurrent load to migrate from.

**Reply by SidB3** [reply]
Thanks for the thorough answer, Ben. I will look at the evictionListener and removalListener
